### DISCUSSION TITLE
pick like osgPick

### DISCUSSION THREAD

**Discussion by bclay1297** [vsg-dev/VulkanSceneGraph#581] [General]
Is there a way to pick / identify nodes like there is with osgPick example?

Bruce

**Comment by rolandhill** [comment]
Hi Bruce,

You're probably looking for the LineSegmentIntersector. Look
at examples/traversals/vsgintersection/

Cheers,
Roland

**Reply by bclay1297** [reply]
Roland:
  Thanks for your reply.  I see a lot of data in that app when I click on the model but I do not see any identifiers to say which model or what part of (articulation part) a model was selected.  i.e. no node name or node id.  Am I missing something?

Bruce

**Comment by rolandhill** [comment]
Hi Bruce,

I haven't looked at the example for a while and don't recall what is
available built-in.

I have a similar need. In my case, I'm generating geometry in real time to
represent various data and the most efficient way I've found is to derive
my own class from LineSegmentIntersector so that I can
calculate intersections rather than rely on testing for intersections with
the underlying geometry. The actual intersection testing is done by a
custom ConstVisitor handler in each object class. Depending on your need,
you might only need the ConstVisitor handler. I might be over-complicating
it, but it achieves the result and you can optimise intersection testing if
you need to do it continuously (ie when moving mouse). It's a work in
progress.

Hope that helps,
Roland


void VisNodeDrillDisc::accept(vsg::ConstVisitor& visitor) const
{
    if (auto v = dynamic_cast<LineIntersector*>(&visitor))
    {
           ..... do your testing here

            auto dpt = vsg::dvec3((double)pt.x, (double)pt.y, (double)pt.z);
            auto is = v->add(dpt, 0.0, {}, 0);   <--- built in intersection
vector
            is->setValue("layer",  myLayer)); <--- your custom intersection
information
            is->setValue("id", myId); <--- your custom intersection
information
    }
    else
    {
        traverse(visitor);
    }

**Reply by bclay1297** [reply]
Roland:
  I tried a similar approach using the existing code with success.

After I loaded the model I set a node name
            model->setValue("modelName", "topModel");

in he callback (intersection I added 
	node->getValue("modelName", modelName);

I noticed multiple nodes are returned in the callback so I have to loop to get to the one that returns this value.

So that gives us a way to differentiate between models in our scene the next step is to find a way to get to nodes within a model such as ailerons on an aircraft model.

Bruce

**Comment by robertosfield** [comment]
Hi Bruce,

On Tue, 15 Nov 2022 at 00:27, bclay1297 ***@***.***> wrote:

> Thanks for your reply. I see a lot of data in that app when I click on the
> model but I do not see any identifiers to say which model or what part of
> (articulation part) a model was selected. i.e. no node name or node id. Am
> I missing something?
>

The LineSegmentIntersector::Intersector structure that is returned for each
hit has a nodePath member:


https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/utils/LineSegmentIntersector.h#L48
